### PR TITLE
Feature is vpnserver

### PR DESCRIPTION
**Pull request by SunithaGudisagarIBM**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVPNServerBasic'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVPNServerRouteBasic'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVPNServersDataSourceBasic'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVPNServerDataSourceBasic'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVPNServerClientConfigDataSourceBasic'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVPNServerRouteDataSourceBasic'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVPNServerRoutesDataSourceBasic'

...
```
Find ref below:
https://github.com/ibm-vpc/terraform-provider-ibm/pull/30#issuecomment-1131431611

### CODE REVIEW

File: ibm/resource_ibm_is_vpn_server.go [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+				Required:    true,
+				ForceNew:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Set:         schema.HashString,
+				Description: "The unique identifier for this subnet. The subnets to provision this VPN server in.  Use subnets in different zones for high availability.",
+			},
+
+			"resource_type": &schema.Schema{
+				Type:         schema.TypeString,
+				Default:      "vpn_server",
```
Comment (ujjwal-ibm) [inline ibm/resource_ibm_is_vpn_server.go:250]: why is this optional ?

File: ibm/resource_ibm_is_vpn_server.go [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+					Schema: map[string]*schema.Schema{
+						"address": &schema.Schema{
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The IP address. This property may add support for IPv6 addresses in the future. When processing a value in this property, verify that the address is in an expected format. If it is not, log an error. Optionally halt processing and surface the error, or bypass the resource on which the unexpected IP address format was encountered.",
+						},
+						"deleted": &schema.Schema{
+							Type:        schema.TypeList,
+							MaxItems:    1,
+							Optional:    true,
```
Comment (ujjwal-ibm) [inline ibm/resource_ibm_is_vpn_server.go:177]: why is this optional ?

File: ibm/resource_ibm_is_vpn_server.go [hunk 3]
```diff
@@ -0,0 +1,806 @@
...
+			"port": &schema.Schema{
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Default:      443,
+				ValidateFunc: InvokeValidator("ibm_is_vpn_server", "port"),
+				Description:  "The port number to use for this VPN server.",
+			},
+			"private_ips": &schema.Schema{
+				Type:        schema.TypeList,
+				Computed:    true,
```
Comment (ujjwal-ibm) [inline ibm/resource_ibm_is_vpn_server.go:165]: nothing inside this should be optional when its computed

File: ibm/sunithatestingfile.txt [hunk 4]
```diff
@@ -0,0 +1,52 @@
+# (C) Copyright IBM Corp. 2021.
+
+client
```
Comment (ujjwal-ibm) [inline ibm/sunithatestingfile.txt:3]: uncheck this file

File: website/docs/d/is_vpn_server.html.markdown [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+Review the argument reference that you can specify for your data source.
+
+- `identifier` - (Optional, String) The ID of the VPN server.
+- `name` - (Optional, String) The name of the VPN server.
+
+## Attribute Reference
+
+In addition to all argument references listed, you can access the following attribute references after your data source is created.
+
+* `identifier` - The unique identifier of the VPNServer.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_vpn_server.html.markdown:31]: update the docs with the standard doc style.
* -> -

File: website/docs/d/is_vpn_server.html.markdown [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+* `certificate` - (Required, List) The certificate instance for this VPN server.
+Nested scheme for **certificate**:
+	* `crn` - (Required, String) The CRN for this certificate instance.
+
+* `client_authentication` - (Required, List) The methods used to authenticate VPN clients to this VPN server. VPN clients must authenticate against all provided methods.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_vpn_server.html.markdown:36]: Remove Required from exported attributes

File: website/docs/d/is_vpn_server_clients.html.markdown [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+Review the argument reference that you can specify for your data source.
+
+* `vpn_server_id` - (Required, Forces new resource, String) The VPN server identifier.
+
+## Attribute Reference
+
+In addition to all argument references listed, you can access the following attribute references after your data source is created.
+
+* `id` - The unique identifier of the VPNServerClientCollection.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_vpn_server_clients.html.markdown:31]: change * to -

File: website/docs/d/is_vpn_server_clients.html.markdown [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	* `remote_port` - (Required, Integer) The remote port of this VPN client.
+	  * Constraints: The maximum value is `65535`. The minimum value is `1`.
+	* `resource_type` - (Required, String) The resource type.
+	  * Constraints: Allowable values are: vpn_server_client
+	* `status` - (Required, String) The status of the VPN client:- `connected`: the VPN client is `connected` to this VPN server.- `disconnected`: the VPN client is `disconnected` from this VPN server.The enumerated values for this property are expected to expand in the future. When processing this property, check for and log unknown values. Optionally halt processing and surface the error, or bypass the VPN client on which the unexpected property value was encountered.
+	  * Constraints: Allowable values are: connected, disconnected
+	* `username` - (Optional, String) The username that this VPN client provided when connecting to the VPN server.This property will be present only when  the`username` client authentication method is enabled on the VPN server.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_vpn_server_clients.html.markdown:56]: remove optional/required from exported attributes

File: website/docs/r/is_vpn_server_route.html.markdown [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+In addition to all argument references listed, you can access the following attribute references after your resource is created.
+
+* `id` - The unique identifier of the VPNServerRoute.
+* `vpn_route` - The identifier of the VPNServerRoute.
+* `created_at` - (Required, String) The date and time that the VPN route was created.
+* `href` - (Required, String) The URL for this VPN route.
+  * Constraints: The value must match regular expression `/^http(s)?:\/\/([^\/?#]*)([^?#]*)(\\?([^#]*))?(#(.*))?$/`
+* `lifecycle_state` - (Required, String) The lifecycle state of the VPN route.
+  * Constraints: Allowable values are: deleting, failed, pending, stable, updating, waiting, suspended
+* `resource_type` - (Required, String) The resource type.
```
Comment (ujjwal-ibm) [inline website/docs/r/is_vpn_server_route.html.markdown:48]: remove optional/required from exported attributes

File: website/docs/r/is_vpn_server_route.html.markdown [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  * Constraints: The value must match regular expression `/^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(\/(3[0-2]|[1-2][0-9]|[0-9]))$/`
+* `name` - (Optional, String) The user-defined name for this VPN route. If unspecified, the name will be a hyphenated list of randomly-selected words. Names must be unique within the VPN server the VPN route resides in.
+  * Constraints: The maximum length is `63` characters. The minimum length is `1` character. The value must match regular expression `/^([a-z]|[a-z][-a-z0-9]*[a-z0-9])$/`
+* `vpn_server_id` - (Required, Forces new resource, String) The VPN server identifier.
+
+## Attribute Reference
+
+In addition to all argument references listed, you can access the following attribute references after your resource is created.
+
+* `id` - The unique identifier of the VPNServerRoute.
```
Comment (ujjwal-ibm) [inline website/docs/r/is_vpn_server_route.html.markdown:41]: change * to -

File: website/docs/r/is_vpn_server.html.markdown [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+## Example Usage
+The following example creates a VPN Server:
+
+```terraform
+resource "ibm_is_vpn_server" "is_vpn_server" {
+  certificate_crn = "crn:v1:staging:public:cloudcerts:us-south:a/efe5afc483594adaa8325e2b4d1290df:86f62739-f3a8-42ac-abea-f23255965983:certificate:00406b5615f95dba9bf7c2ab52bb3083"
+  client_authentication {
+    method    = "certificate"
+    client_ca = "crn:v1:staging:public:cloudcerts:us-south:a/efe5afc483594adaa8325e2b4d1290df:86f62739-f3a8-42ac-abea-f23255965983:certificate:6a85a87d01dd5a4268a8bca16cb998eb"
```
Comment (ujjwal-ibm) [inline website/docs/r/is_vpn_server.html.markdown:21]: remove this information

update all the documents

**Review by ujjwal-ibm — CHANGES_REQUESTED** [review]

File: go.mod [hunk 12]
```diff
@@ -47,3 +47,7 @@ require (
 )
 
 replace github.com/softlayer/softlayer-go v1.0.3 => github.com/IBM-Cloud/softlayer-go v1.0.3-tf
+
+replace github.com/IBM/vpc-go-sdk v0.8.0 => ./common/github.com/IBM/vpc-go-sdk
```
[the remark this answers is not in the document]
Comment (BShrivastav) [reply go.mod]: done.

File: ibm/resource_ibm_is_vpn_server.go (hunk 3 above)
Comment (BShrivastav) [reply ibm/resource_ibm_is_vpn_server.go:165]: done.

File: website/docs/d/is_vpn_server.html.markdown (hunk 5 above)
Comment (BShrivastav) [reply website/docs/d/is_vpn_server.html.markdown:31]: done.

File: website/docs/d/is_vpn_server.html.markdown (hunk 6 above)
Comment (BShrivastav) [reply website/docs/d/is_vpn_server.html.markdown:36]: done.

File: website/docs/r/is_vpn_server_route.html.markdown (hunk 9 above)
Comment (BShrivastav) [reply website/docs/r/is_vpn_server_route.html.markdown:48]: done.

File: website/docs/r/is_vpn_server_route.html.markdown (hunk 10 above)
Comment (BShrivastav) [reply website/docs/r/is_vpn_server_route.html.markdown:41]: done.

File: ibm/data_source_ibm_is_vpn_servers.go [hunk 13]
```diff
@@ -0,0 +1,731 @@
...
+
+	"github.com/IBM/vpc-go-sdk/vpcv1"
+)
+
+func dataSourceIBMIsVPNServers() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceIBMIsVPNServersRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (BShrivastav) [reply ibm/data_source_ibm_is_vpn_servers.go:23]: added.

File: ibm/data_source_ibm_is_vpn_server_routes.go [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+		ReadContext: dataSourceIBMIsVPNServerRoutesRead,
+
+		Schema: map[string]*schema.Schema{
+			"vpn_server": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The VPN server identifier.",
+			},
+			"name": &schema.Schema{
+				Type:        schema.TypeString,
```
[the remark this answers is not in the document]
Comment (SunithaGudisagarIBM) [reply ibm/data_source_ibm_is_vpn_server_routes.go:29]: no

**Comment by SunithaGudisagarIBM** [comment]
Screenshot 2022-05-18 at 4 21 49 PM

Screenshot 2022-05-18 at 4 58 43 PM

Screenshot 2022-05-18 at 6 05 43 PM

Screenshot 2022-05-18 at 6 23 41 PM

Screenshot 2022-05-18 at 6 37 25 PM

Screenshot 2022-05-18 at 6 48 44 PM

Screenshot 2022-05-18 at 7 12 42 PM

Screenshot 2022-05-19 at 2 21 08 PM

File: ibm/service/vpc/resource_ibm_is_vpn_server.go [hunk 15]
```diff
@@ -614,6 +658,39 @@ func resourceIBMIsVPNServerRead(context context.Context, d *schema.ResourceData,
 	if err = d.Set("private_ips", privateIps); err != nil {
 		return diag.FromErr(fmt.Errorf("Error setting private_ips: %s", err))
 	}
+	// if vpnServer.VPC != nil {
```
Comment (SunithaGudisagarIBM) [inline ibm/service/vpc/resource_ibm_is_vpn_server.go:661]: remove unwanted commented code

File: ibm/service/vpc/resource_ibm_is_vpn_server.go [hunk 16]
```diff
@@ -649,6 +726,24 @@ func resourceIBMIsVPNServerReservedIPReferenceDeletedToMap(reservedIPReferenceDe
 	return reservedIPReferenceDeletedMap
 }
 
+func resourceIBMIsVPNServerVPCReferenceToMap(vpcRef vpcv1.VPCReference) map[string]interface{} {
```
Comment (SunithaGudisagarIBM) [inline ibm/service/vpc/resource_ibm_is_vpn_server.go:729]: Remove if its not referenced

**Review by SunithaGudisagarIBM — COMMENTED** [review]
Hey Bhavesh,
I have few comments, please check n incorporate if its relevant

File: website/docs/d/is_vpc_routing_table_route.html.markdown [hunk 17]
```diff
@@ -59,6 +59,22 @@ In addition to all argument references listed, you can access the following attr
 	- `deliver`: deliver the packet to the specified `next_hop`
 	- `drop`: drop the packet.
 - `created_at` - (String) The date and time that the route was created.
+- `creator` - (Optional, List) If present, the resource that created the route. Routes with this property present cannot bedirectly deleted. All routes with an `origin` of `learned` or `service` will have thisproperty set, and future `origin` values may also have this property set.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_vpc_routing_table_route.html.markdown:62]: remove optional in datasource

**Review by ujjwal-ibm — CHANGES_REQUESTED** [review]

File: website/docs/d/is_vpn_server_clients.html.markdown [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+data "ibm_is_vpn_server_clients" "example" {
+	vpn_server = ibm_is_vpn_server.example.id
+}
+```
+
+## Argument Reference
+
+Review the argument reference that you can specify for your data source.
+
+- `vpn_server` - (Required, Forces new resource, String) The VPN server identifier.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_vpn_server_clients.html.markdown:24]: forces new not required in datasource

**Comment by SunithaGudisagarIBM** [comment]
Moving PR to upstream

**Pull request closed without merging**